### ISSUE TITLE
Offline-mode (at least) whitelist is broken

### ISSUE THREAD

**Issue by MithrandirCraft** [PaperMC/Paper#2011]
### What behaviour is expected:
A functional whitelist

### What behaviour is observed:
It's quite probable that you won't be capable of accessing the server even if you are whitelisted. 3rd plugins are necessary.

### Steps/models to reproduce:
Just start up a clean PaperSpigot or Spigot server in offline-mode true and add some players to the whitelist. Then attempt to access the server with one of said playernames, usually through an offline mode compatible client.

### Plugin list:
![imagen](https://user-images.githubusercontent.com/6623704/57465752-8dfb9900-727f-11e9-9370-fae45a994336.png)

### Paper build number:
Paper version git-Paper-624 (MC: 1.13.2) (Implementing API version 1.13.2-R0.1-SNAPSHOT)

**Comment by VADemon** [comment]
Whitelist cannot function in offline-mode anyway.

Online: UUIDs are bound to player acc and verified by session server (in simple terms)

Offline: UUIDs are generated from playername, namely `Offline:<Nickname>`.
As such, any player could rename himself and join the server as the 'whitelisted' person.

If you want to restrict offline-mode access, you either need an auth plugin (either per-player or global password).

If you really need to temporarily close down the server: add another auth plugin (mkay..?) / change port / edit UUIDs in whitelist manually to have Offline-mode UUIDs (not safe as per above)

**Comment by sotr2333** [comment]
Actually, this can be resolved by check online mode in multiple locations(UserCache, etc,.), to apply offline UUID for offline servers instead of lookup an online one.

**Comment by electronicboy** [comment]
![image](https://user-images.githubusercontent.com/1228900/58027890-835bc200-7b11-11e9-9b33-34b155fd6085.png)
I mean, I can't really say that anybody is really interested in delving into offline mode issues, but it appears fine here?

**Comment by Migushthe2nd** [comment]
I have had this issue as well with two players out of the total 14 people who have visited my server. I eventually added them while they were online (with whitelist disabled), and that seemed to have work. The uuid for those two seemed to differ when they were added while not being online.

**Comment by suwermave** [comment]
Migushthe2nd solution confirmed.

**Comment by sgdc3** [comment]
If the player isn't saved in the usercache.json file (he never joined the server before for example) trying to whitelist him will result in whitelisting the ONLINE uuid.

How UUIDs for offline players should be handled instead:

online-mode (server.properties): true
=> resolve online UUID

online-mode: false
bungeecord (spigot.yml): false
=> generate offline UUID from name

online-mode: false
bungeecord: true
bungee-online-mode (paper.yml): true
=> resolve online UUID

online-mode: false
bungeecord: true
bungee-online-mode: false
=> generate offline UUID from name

**Comment by sgdc3** [comment]
This was patched successfully in yatopia: https://github.com/YatopiaMC/Yatopia/blob/ver/1.16.4/patches/server/0030-Use-offline-uuids-if-we-need-to.patch

**Comment by SuspiciousLookingOwl** [comment]
What @sgdc3 said is correct, the (temporary) work around is:

1. Add user to whitelist from console (`whitelist add the_username`)
2.  Go to http://tools.glowingmines.eu/ and enter `the_username`, make sure `Get UUID from player name` is checked
3. Press `Search`, you should see an offline UUID is being generated
![image](https://user-images.githubusercontent.com/32597776/104075143-85254c80-5244-11eb-821e-1cb90d2e8290.png)
4. Open `whitelist.json`
5. Replace the UUID value on `the_username` to the generated offline mode UUID and save the file, e.g.
```json
[
  {
    "uuid": "e2f9a704-9a4d-3ad6-8faa-e48fde4a4809",
    "name": "the_username"
  }
]
```
6. Run `whitelist reload`
7. The player should be able to join now

**Comment by Camotoy** [comment]
Since this issue is updated: should a patch like https://github.com/YatopiaMC/Yatopia/blob/3770a434722cd81011e6045fc7becd50122b03cf/patches/server/0029-Use-offline-uuids-if-we-need-to.patch be brought into Paper, for purely offline mode instances there needs to be a manual override for plugins that spoof the UUID intentionally (for example: https://github.com/YatopiaMC/Yatopia/blob/3770a434722cd81011e6045fc7becd50122b03cf/patches/server/0064-Add-a-special-case-for-floodgate-and-offline-uuids.patch). The latter patch however does not matter under a proxy scenario.

**Comment by imnuke** [comment]
> What @sgdc3 said is correct, the (temporary) work around is:
> 
> 1. Add user to whitelist from console (`whitelist add the_username`)
> 2. Go to http://tools.glowingmines.eu/ and enter `the_username`, make sure `Get UUID from player name` is checked
> 3. Press `Search`, you should see an offline UUID is being generated
>    ![image](https://user-images.githubusercontent.com/32597776/104075143-85254c80-5244-11eb-821e-1cb90d2e8290.png)
> 4. Open `whitelist.json`
> 5. Replace the UUID value on `the_username` to the generated offline mode UUID and save the file, e.g.
> 
> ```json
> [
>   {
>     "uuid": "e2f9a704-9a4d-3ad6-8faa-e48fde4a4809",
>     "name": "the_username"
>   }
> ]
> ```
> 
> 1. Run `whitelist reload`
> 2. The player should be able to join now

worked fine on 1.15.2, but is there any way to do this directly in minecraft console without modifying the file manually?

**Comment by SuspiciousLookingOwl** [comment]
> worked fine on 1.15.2, but is there any way to do this directly in minecraft console without modifying the file manually?

Haven't found easier way to do that. I ended up ditching built-in whitelist system and use [this plugin](https://www.spigotmc.org/resources/nickname-whitelist.40924/) instead, which allows you to whitelist by name. and ignores the UUID.

**Comment by megargayu** [comment]
I'm kind of confused by this issue still being opened. When I use `paper-1.16.5-642.jar`, whitelist works fine for me in offline mode. I tried adding an offline player to a whitelist and I could connect - I tried removing this offline player and I couldn't connect. I did the same to a online player, and the expected behaviour happened. Am I missing something here, or has this bug been fixed?

**Comment by electronicboy** [comment]
On the basis that mojangs servers are having issues, this is likely due to the fact that the names you're testing are already in the user cache, looks like mojang never cared to consider offline mode servers here?

**Comment by electronicboy** [comment]
Just tested with an offline name, works as intended

**Comment by electronicboy** [comment]
Okay, misread the history, vanilla doesn't care to check and likely just fell back to the offline UUID as mojangs API is broken

**Comment by megargayu** [comment]
> Okay, misread the history, vanilla doesn't care to check and likely just fell back to the offline UUID as mojangs API is broken

So, are you saying that once Mojang's API is fixed, the whitelist will stop working again?

**Comment by electronicboy** [comment]
electronicboy closed this in 453e983 13 hours ago

**Comment by megargayu** [comment]
> electronicboy closed this in [453e983](https://github.com/PaperMC/Paper/commit/453e983e4b7546d280c6449a0fbeb676bb3ecc9c) 13 hours ago

OH sorry, I didn't realize you made a commit which fixes this 😅

**Comment by purejosh** [comment]
Appears that this bug is still not fully resolved. Experienced it on 1.18.1 in a velocity environment, on a child server's whitelist. Had to manually add the user's UUID into the whitelist.json correctly, as it was getting the wrong UUID.
